### PR TITLE
[NPUW] Add Mixture of Experts support to model builder

### PR DESCRIPTION
**Pull request by dylanneve1**
### Details:
 - Add Mixture of Experts support to model builder

### Tickets:
 - *EISW-209517*

### AI Assistance:
 - *AI assistance used: yes*
 - *Claude Code used for coding assistance and to analyze GPT OSS model to recreate patterns, manually tested and verified*

Correct subgraphs produced

Generate:
```
{
  "pipeline_name": "npuw_synthetic_decoder_kv1152",
  "repeated_numbers": {
    "total_subgraphs": 1,
    "synthetic_decoder_kv1152_0_FCEW000": 1
  },
  "subgraphs": [
    "synthetic_decoder_kv1152_0_FCEW000.xml"
  ]
}
```
Prefill:
```
{
  "pipeline_name": "npuw_synthetic_decoder_prefill",
  "repeated_numbers": {
    "total_subgraphs": 5,
    "synthetic_decoder_prefill_00_FCEW000": 1,
    "synthetic_decoder_prefill_01_REP009C": 12,
    "synthetic_decoder_prefill_02_REP00B7": 12,
    "synthetic_decoder_prefill_03_REP00B9": 11,
    "synthetic_decoder_prefill_36_FCEW001": 1
  },
  "subgraphs": [
    "synthetic_decoder_prefill_00_FCEW000.xml",
    "synthetic_decoder_prefill_01_REP009C.xml",
    "synthetic_decoder_prefill_03_REP00B9.xml",
    "synthetic_decoder_prefill_36_FCEW001.xml"
  ],
  "moe_subgraphs": [
    "synthetic_decoder_prefill_02_REP00B7_moe_chunk_16.xml",
    "synthetic_decoder_prefill_02_REP00B7_moe_chunk_32.xml",
    "synthetic_decoder_prefill_02_REP00B7_moe_chunk_64.xml",
    "synthetic_decoder_prefill_02_REP00B7_moe_chunk_128.xml",
    "synthetic_decoder_prefill_02_REP00B7_moe_chunk_256.xml"
  ]

```

### CODE REVIEW

File: src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp [hunk 1]
```diff
@@ -824,6 +856,257 @@ ov::Output<ov::Node> GELU::operator()(const ov::Output<ov::Node>& input, const s
...
+    auto expert_mask = std::make_shared<ov::opset11::Transpose>(one_hot, transpose_order);
+    expert_mask->set_friendly_name(name + ".expert_mask");
+
+    // --- Prepare hidden states: Convert -> Unsqueeze -> Reshape ---
+    auto hidden_cvt = std::make_shared<ov::opset11::Convert>(input_2d, precision);
+    hidden_cvt->set_friendly_name(name + ".hidden_cvt");
+
+    auto target_shape = std::make_shared<ov::opset11::ShapeOf>(hidden_cvt, ov::element::i64);
+    target_shape->set_friendly_name(name + ".target_shape");
+
```
[the remark this answers is not in the document]
Comment (dylanneve1) [reply src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp]: Fixed — removed the unused `target_shape` variable. `target_shape_i32` (i32 variant) is the one actually used for the accumulator broadcast.

File: src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp [hunk 2]
```diff
@@ -1424,8 +1707,21 @@ std::shared_ptr<ov::Model> ModelBuilder::build_llm(const LLMConfig& config_in) {
...
-    if (!config.ffn)
-        config.ffn = SwiGLU(config.hidden_size, config.intermediate_size, config.precision, config.weight);
+    if (!config.ffn) {
+        if (config.num_experts > 0) {
+            size_t moe_inter = config.moe_intermediate_size > 0
+                                   ? config.moe_intermediate_size
+                                   : config.intermediate_size;
+            size_t moe_k = config.num_experts_per_tok > 0
+                               ? config.num_experts_per_tok
+                               : 2;
```
[the remark this answers is not in the document]
Comment (dylanneve1) [reply src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp]: Fixed — added `std::min<size_t>(2, config.num_experts)` for the default and an `OPENVINO_ASSERT` validating `1 <= moe_k <= num_experts`.

File: src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp [hunk 3]
```diff
@@ -1424,8 +1707,21 @@ std::shared_ptr<ov::Model> ModelBuilder::build_llm(const LLMConfig& config_in) {
...
+            config.ffn = MoEFFN(config.hidden_size, moe_inter, config.num_experts,
+                                moe_k, config.precision, config.weight,
+                                config.shared_expert_ffn);
+        } else {
+            config.ffn = SwiGLU(config.hidden_size, config.intermediate_size, config.precision, config.weight);
+        }
```
[the remark this answers is not in the document]
Comment (dylanneve1) [reply src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp:1621]: Added two E2E tests in `moe_transformation_test.cpp`:
- `BuildLLM_GPTOSSBatchedMoE_AnalyzeStructure` — builds LLM with `GPTOSSBatchedMoEFFN`, verifies Tile/TopK/ReduceSum exist and `analyze_moe_structure` succeeds
- `BuildLLM_GPTOSSBatchedMoE_RouterPattern` — verifies router TopK(MAX) naming and ScatterElementsUpdate per layer

File: src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp [hunk 4]
```diff
@@ -824,6 +856,257 @@ ov::Output<ov::Node> GELU::operator()(const ov::Output<ov::Node>& input, const s
...
+        auto accum_slice = std::make_shared<ov::opset11::Slice>(accumulator,
+                                                                accum_slice_start, accum_slice_stop,
+                                                                accum_slice_step, accum_slice_axes);
+        accum_slice->set_friendly_name(en + ".accum_slice");
+
+        auto accum_slice_shape = std::make_shared<ov::opset11::ShapeOf>(accum_slice, ov::element::i32);
+        accum_slice_shape->set_friendly_name(en + ".accum_slice_shape");
+
+        auto scatter_idx_broadcast = std::make_shared<ov::op::v3::Broadcast>(
+            scatter_idx_2d, accum_slice_shape, ov::op::BroadcastType::BIDIRECTIONAL);
```
[the remark this answers is not in the document]
Comment (dylanneve1) [reply src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp]: The Slice is intentional and correct. It creates a `[1, hidden]` shape used for **bidirectional** broadcast with `[active, 1]`, giving `[active, hidden]`. Without the Slice, using the full accumulator shape `[tokens, hidden]` would broadcast the index to `[tokens, hidden]` instead of `[active, hidden]`, which is wrong — we only want to scatter the active tokens' data, not all tokens.

The ScatterElementsUpdate then correctly writes `[active, hidden]` updates into the `[tokens, hidden]` accumulator at the positions specified by `token_indices`.

File: src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp [hunk 5]
```diff
@@ -824,6 +856,398 @@ ov::Output<ov::Node> GELU::operator()(const ov::Output<ov::Node>& input, const s
...
+    tile_repeats = C::create(ov::element::i64, ov::Shape{2}, std::vector<int64_t>{ne_i, 1});
+    slice_step = C::create(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{1});
+    slice_axis2 = C::create(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{2});
+    slice_start_0 = C::create(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{0});
+    slice_stop_is = C::create(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{is_i});
+    slice_start_is = C::create(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{is_i});
+    slice_stop_2is = C::create(ov::element::i64, ov::Shape{1}, std::vector<int64_t>{2 * is_i});
+    min_const = C::create(ov::element::f32, ov::Shape{1}, std::vector<float>{20.0f});
+    swish_beta = C::create(ov::element::f32, ov::Shape{}, std::vector<float>{1.0f});
+    clamp_add_zero = C::create(ov::element::f32, ov::Shape{1}, std::vector<float>{0.0f});
```
[the remark this answers is not in the document]
Comment (dylanneve1) [reply src/plugins/intel_npu/tests/functional/behavior/npuw/test_engine/models/model_builder.cpp]: These f32 constants match what GPT-OSS 20B produces — the original PyTorch model exports `aten::clamp` constants as f32 regardless of weight precision. OpenVINO's Minimum, Swish, and Add ops handle mixed precision through implicit type promotion.

However, to be safe and match the suggestion, I'll make them use `precision` since we control the model builder and there's no reason to force f32.

File: src/plugins/intel_npu/tests/unit/npuw/moe_transformation_test.cpp [hunk 6]
```diff
@@ -557,4 +559,39 @@ TEST_F(MoETransformationTest, AWQvsNonAWQComparison) {
...
+        }
+        if (std::dynamic_pointer_cast<ov::op::v3::ScatterElementsUpdate>(op))
+            scatter_count++;
+    }
+
+    EXPECT_TRUE(has_tile) << "Missing Tile (expert)";
+    EXPECT_TRUE(has_topk) << "Missing TopK (router)";
+    EXPECT_TRUE(has_reduce_sum) << "Missing ReduceSum (aggregation)";
+    EXPECT_EQ(topk_router_count, 2u) << "One router TopK per layer";
+    EXPECT_EQ(scatter_count, 2u) << "One ScatterElementsUpdate per layer";
```
Comment (intelgaoxiong) [inline src/plugins/intel_npu/tests/unit/npuw/moe_transformation_test.cpp:601]: Softmax is a key node.
It's better to check it too.

**Review by intelgaoxiong — APPROVED** [review]
LGTM

**Review by dmatveev — APPROVED** [review]
